src/merkle.ts

```diff
@@ -19,49 +18,53 @@ function padLeaves(hashes: Hash[]) {
   return [...hashes, ...padding];
 }
 
-/**
- * Given a merkle proof for a binary tree, computes the index of the node for the proof.
- * @param proof A list of siblings ordered from bottom of the tree to the top.
- * @returns Index of the node.
- */
-export function proofToIndex(proof: Proof): number {
-  let index = 0;
-  let pow = 0;
-  for (const sibling of proof) {
-    if ('left' in sibling) {
-      index += Math.pow(2, pow);
-    }
-    pow += 1;
-  }
-  return index;
-}
-
 export function generateWitness(hashes: Hash[], index: number): WitnessProof {
   const tree = new MerkleTree({hashType: 'SHA3-256'});
   tree.addLeaves(padLeaves(hashes), false);
   tree.makeTree();
 
   const root = tree.getMerkleRoot()?.toString('hex');
   const witness = tree.getLeaf(index)?.toString('hex');
-  const proof = tree.getProof(index);
+  const proof = tree.getProof(index, false);
 
   if (!root || !witness || !proof) {
     throw new Error('Invalid node or proof!');
   }
-  return {witness, proof};
+
+  const nodes = proof.map(p => ('left' in p ? p.left : p.right));
+
+  return {witness, nodes, index};
 }
 
 export function validateWitness(witnessProof: WitnessProof, root: string, depth: number): boolean {
   const tree = new MerkleTree({hashType: 'SHA3-256'});
-  const {proof, witness} = witnessProof;
+  const {index, nodes, witness} = witnessProof;
 
-  if (witnessProof.proof.length !== depth) {
+  if (witnessProof.nodes.length !== depth) {
     throw new Error(
-      `The witness provided is not for a leaf node. Expected ${depth} witness length, recieved ${witnessProof.proof.length}`
+      `The witness provided is not for a leaf node. Expected ${depth} witness length, recieved ${witnessProof.nodes.length}`
     );
   }
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  return tree.validateProof(proof as any, witness, root);
+
+  const convertedProof = generateMerkleToolsProof(nodes, index);
+  return tree.validateProof(convertedProof, witness, root);
+}
+
+function generateMerkleToolsProof(nodes: Hash[], index: number): MerkleToolsProof<string> {
+  const merkleToolsProof: MerkleToolsProof<string>[] = [];
+
+  for (let i = 0; i < nodes.length; i++) {
+    if (index % 2 !== 0) {
+      merkleToolsProof.push({left: nodes[i]});
+    } else {
+      merkleToolsProof.push({right: nodes[i]});
+    }
+
+    index = Math.floor(index / 2);
+  }
+
+  // TODO: The MerkleTools library isn't typed properly so we force a cast here
```

Comment:
Not sure I fully understand. It seems to me that an array is being cast to a non-array object? Is that the right thing to do? It seems that we are casting `MerkleToolsProof<string>[]` to `MerkleToolsProof<string>`

Reply:
Maybe a better approach is to use type assertion for `validateProof`. It seems like converting the return object to the incorrect type here could lead to misuse in other places if this return object is reused.

Reply:
I've moved the cast and todo to the call of  `validateProof` in 4fb7b19